### PR TITLE
Support `Host` header overriding URI

### PR DESCRIPTION
**Pull request by badslug**
This PR adds the ability to set the `Host` header of a request in order to use it for the signature calculation even if the URI host is different. This allows sending requests to the AWS endpoints using SSH tunnels (and presumably other proxies) while calculating the expected AWS signature.

This resolves issue #31

### CODE REVIEW

**Comment by riy** [comment]
There still seems to be an issue.

I installed the plugin as per your instructions and the version has changed to 1.1.0:

image

I put in my credentials...
image

... and configured the `X-Amz-Date`:
image

When I put in the `host` value like this...
image

... and called my SSH tunnel URL `https://localhost:1443` via `GET` I got this message:
```html
HTTP/1.1 400 Bad Request
Server: awselb/2.0
Date: Fri, 13 May 2022 15:49:17 GMT
Content-Type: text/html
Content-Length: 122
Connection: close

<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
</body>
</html>
```

When I remove the `https` from the `host`, just for testing purposes, I get this:
```html
HTTP/1.1 403 Forbidden
Date: Fri, 13 May 2022 15:55:49 GMT
Content-Type: application/json
Content-Length: 192
Connection: close
x-amzn-RequestId: 4e6axxxx-xxxx-xxxx-xxxx-xxxx307416c3
Access-Control-Allow-Origin: *

{"message":"The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details."}
```

I triple-checked if the correct (new) version is installed by looking into the Extensions directory at `/Users/<username>/Library/Containers/com.luckymarmot.Paw/Data/Library/Application Support/com.luckymarmot.Paw/Extensions` and I see the changes in line #256.

Are there more factors that are taken into consideration for the signature calculation?

**Comment by badslug** [comment]
@riy the host header should just be the domain name without the scheme or port. So just vpc-opensearch-...com

Let me know if that works.

**Comment by riy** [comment]
@badslug

Yes, that's what I tried when I removed the `https` scheme from the `host`. And that's when I got the `The request signature we calculated does not match the signature you provided [...]` message:

image

I used copy/paste for everything and checked for whitespaces, invisible characters etc. on both sides.

**Comment by badslug** [comment]
Ok. Thanks. Let me review and get back to you.

**Comment by riy** [comment]
One more thing that I just tried:
I thought maybe the signature generation has a problem with how it's handling special characters in the access credentials. So I kept generating new access keys until I ended up with a secret access key that doesn't contain any slashes, plus-signs etc., just lower- and uppercase letters and digits. But I'm still getting the signature calculation error message.

File: AWSSignature4DynamicValue.js [hunk 1]
```diff
@@ -253,7 +253,13 @@ var AWSSignature4DynamicValue = function() {
...
         }
 
-        headers['host'] = uri.host.toLowerCase();
+        if (headers.host) {
+            // Use the host header if set
+            headers.host = headers.host.toLowerCase();
+        } else {
+            // Otherwise, extract it from the target uri
+            headers.host = uri.host.toLowerCase();
+        }
```
Comment (riy) [inline AWSSignature4DynamicValue.js:262]: Github doesn't let me comment down there, but the `target` variable in line [#278](https://github.com/badslug/Paw-AWSSignature4DynamicValue/pull/33/files#diff-0ee7f581bffa4b47df15029f6826c7d7bae81f9d2bf9407e7b405fa0f9c6e9e8R278) is never used.

Comment (badslug) [reply AWSSignature4DynamicValue.js:262]: Thanks, that's just unused code from an earlier version. I'll remove it on the next commit.

**Comment by riy** [comment]
Hey @badslug,

I used this PR as a base and changed two lines in my PR #34. Please let me know what you think.

**Pull request merged**